mvpoly/src/monomials.rs

```diff
@@ -0,0 +1,480 @@
+use ark_ff::{One, PrimeField, Zero};
+use rand::RngCore;
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    ops::{Add, Mul, Neg, Sub},
+};
+
+use crate::{
+    prime,
+    utils::{naive_prime_factors, PrimeNumberGenerator},
+};
+
+/// Represents a multivariate polynomial in `N` variables with coefficients in
+/// `F`. The polynomial is represented as a sparse polynomial, where each
+/// monomial is represented by a vector of `N` exponents.
+// We could use u8 instead of usize for the exponents
+// FIXME: the maximum degree D is encoded in the type to match the type
+// prime::Dense
+#[derive(Clone)]
+pub struct Sparse<F: PrimeField, const N: usize, const D: usize> {
+    pub monomials: HashMap<[usize; N], F>,
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Add for Sparse<F, N, D> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let mut monomials = self.monomials.clone();
+        for (exponents, coeff) in other.monomials {
+            monomials
+                .entry(exponents)
+                .and_modify(|c| *c += coeff)
+                .or_insert(coeff);
+        }
+        // Remove monomials with zero coefficients
+        let monomials: HashMap<[usize; N], F> = monomials
+            .into_iter()
+            .filter(|(_, coeff)| !coeff.is_zero())
+            .collect();
+        // Handle the case where the result is zero because we want a unique
+        // representation
+        if monomials.is_empty() {
+            Self::zero()
+        } else {
+            Sparse::<F, N, D> { monomials }
+        }
+    }
+}
+
+impl<const N: usize, const D: usize, F: PrimeField> Add<&Sparse<F, N, D>> for Sparse<F, N, D> {
+    type Output = Sparse<F, N, D>;
+
+    fn add(self, other: &Sparse<F, N, D>) -> Self::Output {
```

Comment:
isn't the same function implemented l28 ?

Reply:
:sweat_smile: 
We can use macros, but it is not really readable.
There might also be a better way that I don't know.